### DISCUSSION TITLE
flask integration - Unable to serve more than one web requests without restarting server

### DISCUSSION THREAD

**Discussion by prachijain78** [sqlalchemy/sqlalchemy#7279] [Usage Questions]
Hi, I have been trying to write a basic web application using flask and sqlalchemy. After reading a lot of pros and cons about flask-sqlalchemy, i decided not to go with it and am integrating flask with sqlalchemy on my own.
I am using a 4D database with pyodbc dialect. There is no connection pooling in my dialect and its very similar to an access dialect. I think the dialect part is fine, but where I am stuck is when I try to integrate SQLAlchemy with Flask. I am using reflection to access the tables in my 4D database. The first web request works fine for me and I see the data being fetched from database but any subsequent request fails. Debugging shows that the code fails in inspection.py. It looks like that the engine object (subject) is not in the _registrars list anymore, at least that is my interpretation of the issue. I am not sure why should that happen. I would really appreciate if anyone can help me out here, its been many days and I have tried writing this code in many different ways but I always end up stuck at the same place. On a side note even with using Flask-SQLAlchemy, i get stuck at the same place.

My code looks like below. I have 2 files - 4dengine.py and app.py

in 4dengine.py code is
```py
engine = create_engine('ENGINE_URL', echo=True, echo_pool=True, hide_parameters=False,
                       module=None, paramstyle="qmark", poolclass=NullPool) 
Base = declarative_base() 
metadata = MetaData(bind=engine) 
app_session = sessionmaker(bind=engine, autoflush=False, expire_on_commit=False) 
```
I do not think I need the 'Base' here though

in app.py, the code is
```py
necessary imports from 4dengine and others...
...
app = Flask(__name__)
...
@app.route('/getall')
def getAllFaculty():
     try:
            db_session = scoped_session(app_session)
            session = db_session()
            Faculty = Table('Faculty', metadata, autoload=True, autoload_with=engine)
            result = session.query(Faculty).all()
            for row in result:
                print('ROW --> ', row)
    except Exception as e:
        raise e
    finally:
        db_session.commit();
        db_session.remove()
...
@app.route('/getstudents')
def getAllStudents():
 # this method is exactly the same as getAllFaculty except that it fetches data from a different table.
```
I have tried many combinations like having only one scoped_session and then instantiating the session in web requests, that is
have this line ` db_session = scoped_session(app_session)` right after the app declaration and not in individual requests.
I also tried with not committing or removing the session. (I think commit is not required anyways, since I am just doing fetch operations here)

If in the browser i go to the request `getAll `(that is for getAllFaculty) it works fine and i get all the faculty data but if i refresh that page or instead try the request `getstudents` the code just exits and the server stops. 
Code exits at this line in sqlalchemy/inspection.py
`            ret = reg(subject)
`
and i have the following variable values in this module:

```py
subject -- Engine(4d+pyodbc://@4d)
type_ -->  <class 'sqlalchemy.engine.base.Engine'>
cls -->  <class 'sqlalchemy.engine.base.Engine'>
_registrars -->  defaultdict(<class 'list'>, {<class 'sqlalchemy.sql.elements.ClauseElement'>: True, <class 'sqlalchemy.sql.schema.SchemaItem'>: True, <class 'sqlalchemy.sql.schema.FetchedValue'>: True, <class 'sqlalchemy.sql.lambdas.LambdaElement'>: <function insp at 0x000001B1B195C430>, <class 'sqlalchemy.engine.interfaces.Connectable'>: <function Inspector._connectable_insp at 0x000001B1B1A44DC0>, <class 'sqlalchemy.engine.base.Engine'>: <function Inspector._engine_insp at 0x000001B1B1A44F70>, <class 'sqlalchemy.engine.base.Connection'>: <function Inspector._connection_insp at 0x000001B1B1A4C040>, <class 'sqlalchemy.engine.reflection.Inspector'>: True, <class 'object'>: <function _inspect_mapped_object at 0x000001B1B22D2E50>, <class 'type'>: <function _inspect_mapped_class at 0x000001B1B22EA280>, <class 'sqlalchemy.orm.interfaces.MapperProperty'>: True, <class 'sqlalchemy.orm.interfaces.PropComparator'>: True, <class 'sqlalchemy.orm.attributes.QueryableAttribute'>: True, <class 'sqlalchemy.orm.state.InstanceState'>: True, <class 'sqlalchemy.orm.util.AliasedClass'>: <function <lambda> at 0x000001B1B2372C10>, <class 'sqlalchemy.orm.util.AliasedInsp'>: <function <lambda> at 0x000001B1B2372CA0>, <class 'sqlalchemy.orm.util.Bundle'>: True, <class 'sqlalchemy.orm.mapper.Mapper'>: True, <class 'sqlalchemy.orm.clsregistry._GetColumns'>: <function <lambda> at 0x000001B1B25411F0>, <class 'sqlalchemy.orm.decl_api.DeclarativeMeta'>: <function _inspect_decl_meta at 0x000001B1B254CE50>, <class 'sqlalchemy.orm.query.Query'>: True})
reg -->  <function Inspector._engine_insp at 0x000001B1B1A44F70>
```

I think I might be missing something very basic here. I am new to SQLAlchemy and Flask, and would really appreciate all help. Thanks.

**Comment by CaselIT** [comment]
Hi,

First of all, `bind` in metadata is deprecated. (It should not cause this issue tough)
You should probably avoid reflecting the table at each request, probably should do that on module load instead (or on the first request).

Also looking at how you are using the session you don't need to create a scoped session, you can just get a session by calling app_session.

Have you tried to just do multiple queries without flask?
Like something like this work?
```py
engine = create_engine('ENGINE_URL', echo=True, echo_pool=True, hide_parameters=False,
                       module=None, paramstyle="qmark", poolclass=NullPool) 
Base = declarative_base() 
metadata = MetaData() 
app_session = sessionmaker(bind=engine, autoflush=False, expire_on_commit=False) 
Faculty = Table('Faculty', metadata, autoload=True, autoload_with=engine)

def go():
     try:
            session = app_session()
            result = session.query(Faculty).all()
            for row in result:
                print('ROW --> ', row)
    finally:
        db_session.commit();
        db_session.remove()

for _ in range(5):
    go()
```

**Reply by prachijain78** [reply]
Thank you for your prompt reply and the feedback on the code. I used scoped_session as I read in various forums that that is the way to go if we are expecting multiple threads or if its a web application and as that is something the flask-sqlalchemy package also does. But  I am happy to get rid of it if it is not really needed. Monetheless, code does not work with it or without it, either way. I had already tried without flask and it works fine. In fact, with flask too if I do multiple queries in one web request it seems to work alright. The issue comes only when I try multiple web requests.
About reflection, I did try not reflecting the table every request but then the request does not work even once. If i have this line
`Faculty = Table('Faculty', metadata, autoload=True, autoload_with=engine)` outside the app or right after the app declaration, and I try to use `Faculty` like a class, it does not seem to work. In such a case when i do the query - `result = session.query(Faculty).all()`, it does not seem to have an iterator object. I think I cannot treat this as a class unless i explicitly declare it to be a class. 
It seems like once reflection is done, requests do not work.  I am trying to work off of a very old legacy database that has about 200 tables of which i only need about 5-10 tables, hence, I cannot reflect the entire database also, and each of these tables that I need, have about 50-60 columns.
I am happy to try any more alternatives or if you have any more pointers. I am not sure why only one request works and that too only if i reflect the related table in that request.

**Reply by CaselIT** [reply]
Regarding scoped session it's not that's not needed, my suggestion was based on how you create it. If you want to use use it you should create it once at the module level and then call on it as if it was a normal session instance. But to properly integrate it with flask you should also remove it after the request has finished. For your example it's probably fine to just use the session maker

>  I had already tried without flask and it works fine. 

did you try the example above?

> In fact, with flask too if I do multiple queries in one web request it seems to work alright. The issue comes only when I try multiple web requests.

The request are one sequential or at the same time? do you have an error?

> About reflection, I did try not reflecting the table every request but then the request does not work even once. If i have this line
Faculty = Table('Faculty', metadata, autoload=True, autoload_with=engine) outside the app or right after the app declaration, and I try to use Faculty like a class, it does not seem to work.

I find this very strange. what is the error?

**Reply by prachijain78** [reply]
> 
> Regarding scoped session it's not that's not needed, my suggestion was based on how you create it. If you want to use use it you should create it once at the module level and then call on it as if it was a normal session instance. But to properly integrate it with flask you should also remove it after the request has finished. For your example it's probably fine to just use the session maker
> 
Understand and yes that is logical too.

> > I had already tried without flask and it works fine.
> 
> did you try the example above?
> 
Yes, and it does work fine. I could print all the rows 5 times.

> > In fact, with flask too if I do multiple queries in one web request it seems to work alright. The issue comes only when I try multiple web requests.
> 
> The request are one sequential or at the same time? do you have an error?
> 

sequential requests.
I do not have an error. But i had put print statements in the code for sqlalchemy package to know until which point the request works. Looks like it breaks in inspection. It prints everything before line 64 in inspection.py, that is, 
` ret = reg(subject)` but none of my print statements get executed after this line, code just exits without an error. So i am guessing that the code breaks when it tries to reflect the table again in another request or when it tries to reflect any other table in another request.

> > About reflection, I did try not reflecting the table every request but then the request does not work even once. If i have this line
> > Faculty = Table('Faculty', metadata, autoload=True, autoload_with=engine) outside the app or right after the app declaration, and I try to use Faculty like a class, it does not seem to work.
> 
> I find this very strange. what is the error?

Again, I did not have an error. But the code stopped at `fn()` call in `_wrap_pool_connect()` in sqlalchemy.engine,base

Well, this got me to think that i have an issue with connection pooling since only the first request works and none other. Since 4D does not support connection pooling, I had set the poolclass to NullPool and this was causing the issue with reflecting the table outside the requests. I removed the NullPool class and left it to default QueuePool, after which I could have the first request working with the tables reflected outside the app but second request still did not work. I think since 4D does not support pooling, I had these issues. I finally tried the `StaticPool` class and looks like it works as it maintains a single connection across all requests. I do not know if this will cause any issues in the future when we have multiple connections etc but for now this seems to be working.

Thank you for all your help and guidance. This was extremely helpful to point me in the right direction.

**Comment by prachijain78** [comment]
Needed to switch from 'NullPool' to 'StaticPool' as poolclass in the dialect to be able to get multiple requests working with one connection.
